michelson.md

```diff
@@ -1790,6 +1857,79 @@ identical to those defined over integers.
     requires I2 >Int 0
 ```
 
+### Ticket operations
+
+```k
+  rule <k> TICKET _A => . ...</k>
+       <stack> [T X] ;
+               [nat N] ;
+               SS
+            => [(ticket T) #Ticket(Addr, X, N)] ;
+               SS
+       </stack>
+       <myaddr> Addr </myaddr>
+     requires #IsComparable(T)
+
+  rule <k> READ_TICKET _A => . ...</k>
+       <stack> [(ticket T) #Ticket(Addr, X, N)] ;
+               SS
+            => [(pair address (pair T nat)) (Pair Addr (Pair X N))] ;
+               [(ticket T) #Ticket(Addr, X, N)] ;
+               SS
+       </stack>
+
+  rule <k> JOIN_TICKETS _A => . ...</k>
+       <stack> [(pair (ticket CTy) (ticket CTy))
+                (Pair #Ticket(S, X, N1) #Ticket(S, X, N2))
+               ] ;
+               SS
+            => [(option (ticket CTy)) (Some #Ticket(S, X, (N1 +Int N2)))] ;
+               SS
+       </stack>
+
+  rule <k> JOIN_TICKETS _A => . ...</k>
+       <stack> [(pair (ticket CTy) (ticket CTy))
+                (Pair #Ticket(#Address(S1), _X1, _N1) #Ticket(#Address(S2), _X2, _N2))
+               ] ;
+               SS
+            => [(option (ticket CTy)) None] ;
+               SS
+       </stack>
+     requires notBool S1 ==String S2
+
+  rule <k> JOIN_TICKETS _A => . ...</k>
+       <stack> [(pair (ticket CTy) (ticket CTy))
+                (Pair #Ticket(_S1, X1, _N1) #Ticket(_S2, X2, _N2))
+               ] ;
+               SS
+            => [(option (ticket CTy)) None] ;
+               SS
+       </stack>
+     requires notBool X1 ==K X2
+      andBool isValue(CTy, X1)
+      andBool isValue(CTy, X2)
```

Comment:
It should be impossible to construct an ill-typed ticket using the external representation, so let's not worry about the checking the `isValue` condition. Given that both rules have the same lefthand side, it would be nicer if we could collapse them.

```suggestion
  rule <k> JOIN_TICKETS _A => . ...</k>
       <stack> [(pair (ticket CTy) (ticket CTy))
                (Pair #Ticket(#Address(S1), X1, _N1) #Ticket(#Address(S2), X2, _N2))
               ] ;
               SS
            => [(option (ticket CTy)) None] ;
               SS
       </stack>
     requires S1 =/=String S2
       orBool X1 =/=K X2
```